File: solutions_2453486_1/Python/Idolstarastronomer/tictac.py
#!/usr/bin/env python
import sys
import logging

def main():
    import argparse
    parser = argparse.ArgumentParser(description='')
    parser.add_argument('input_file',help='The program\'s input file.')
    parser.add_argument('-o','--outfile',required=False,help='The program\'s output file.')
    parser.add_argument('-d','--debug',help='Debug logging level',action="store_true")
    
    args = parser.parse_args()

    if args.debug:
        logging.basicConfig(level=logging.DEBUG)

    f = open(args.input_file,'r')
    fout = sys.stdout
    if args.outfile:
        fout = open(args.outfile,'w')

    cases = int(f.readline())

    for case_number in range(cases):
        logging.debug('Case %d',case_number+1)
        board = []
        for i in range(4):
            board.extend(f.readline().strip())
        f.readline()

        win = test_horizontal_win(board)
        complete = False
        if not win:
            win = test_vertical_win(board)
        if not win:
            win = test_diagonal_win(board)
        if not win:
            complete = test_finished(board)

        if win:
            phrase = '%s won'%win
        else:
            if complete:
                phrase = 'Draw'
            else:
                phrase = 'Game has not completed'

        print >>fout,'Case #%d: %s'%(case_number+1,phrase)

def test_finished(board):
    if '.' in board:
        return False
    else:
        return True

def test_win(row):
    if '.' not in row:
        if 'X' not in row:
            return 'O'
        if 'O' not in row:
            return 'X'

    return None

def test_horizontal_win(board):
    win = None
    for i in range(4):
        win = test_win(board[4*i:4*i+4])
        if win:
            break

    return win

def test_vertical_win(board):
    win = None
    for i in range(4):
        win = test_win(board[i:16:4])
        if win:
            break

    return win

def test_diagonal_win(board):
    win = test_win(board[0:16:5])
    if not win:
        win = test_win(board[3:13:3])

    return win


if __name__=='__main__':
    main()

File: solutions_2453486_1/C++/blackkukreja/tictac.cpp
#include<iostream>
#include<cstdio>
using namespace std;

int main()  {
    int T, i, j, k;
    char board[10][10], win;
    int pl_x, pl_y;
    bool draw;

    freopen("A-large.in", "r", stdin);
    freopen("A-large.txt", "w", stdout);

    cin >> T;
    for(i=1; i<=T; i++) {
        for(j=0; j<4; j++)
            scanf("%s", board[j]);

        win = '.';
        for(j=0; j<4; j++)  {
            pl_x = pl_y = 0;
            for(k=0; k<4; k++)  {
                switch(board[j][k]) {
                    case 'X': pl_x++; break;
                    case 'O': pl_y++; break;
                    case 'T': pl_x++; pl_y++; break;
                }
            }
            if(pl_x == 4)   { win = 'X'; break; }
            else if(pl_y == 4)  { win = 'O'; break; }

            pl_x = pl_y = 0;
            for(k=0; k<4; k++)  {
                switch(board[k][j]) {
                    case 'X': pl_x++; break;
                    case 'O': pl_y++; break;
                    case 'T': pl_x++; pl_y++; break;
                }
            }
            if(pl_x == 4)   { win = 'X'; break; }
            else if(pl_y == 4)  { win = 'O'; break; }
        }
        if(win == '.')  {
            pl_x = pl_y = 0;
            for(j=0; j<4; j++)  {
                switch(board[j][j])   {
                    case 'X': pl_x++; break;
                    case 'O': pl_y++; break;
                    case 'T': pl_x++; pl_y++; break;
                }
            }
            if(pl_x == 4)   { win = 'X'; }
            else if(pl_y == 4)  { win = 'O'; }
        }
        if(win == '.')  {
            pl_x = pl_y = 0;
            for(j=0; j<4; j++)  {
                switch(board[j][3-j])   {
                    case 'X': pl_x++; break;
                    case 'O': pl_y++; break;
                    case 'T': pl_x++; pl_y++; break;
                }
            }
            if(pl_x == 4)   { win = 'X'; }
            else if(pl_y == 4)  { win = 'O'; }
        }
        if(win == 'X')  cout << "Case #" << i << ": " << "X won\n";
        else if(win == 'O') cout << "Case #" << i << ": " << "O won\n";
        else    {
            draw = true;
            for(j=0; draw && j<4; j++)
                for(k=0; draw && k<4; k++)
                    if(board[j][k] == '.')
                        draw = false;
            if(draw)    cout << "Case #" << i << ": "  << "Draw\n";
            else        cout << "Case #" << i << ": "  << "Game has not completed\n";
        }
    }

    return 0;
}
